### PR TITLE
Clarify the optionality of the cnf claim

### LINKED ISSUE #196: Consider making cnf optional
Not all VCs require key binding. For those, cnf should be made optional.

### PR DESCRIPTION
**Pull request by bc-pi**
Clarify the optionality of the cnf claim (to fix #196)

### CODE REVIEW

**Review by Sakurann — APPROVED** [review]

**Review by Sakurann — CHANGES_REQUESTED** [review]
Even if there might be a difference in how a key inside cnf claim is represented (jwk, kid, etc) it should still be cnf claim for the interop purposes, no? What are you trying to account for by making cnf optional?

I think it should remain "required when cryptographic binding is used"

**Comment by bc-pi** [comment]
> Hang on. Even if there might be a difference in how a key inside cnf claim is represented (jwk, kid, etc) it should still be cnf claim for the interop purposes, no? What are you trying to account for by making cnf optional?

please see https://github.com/oauth-wg/oauth-sd-jwt-vc/pull/213

**Comment by Sakurann** [comment]
@bc-pi you probably made the comment while I was updating my original comment based on re-reading the PR. I still think it should remain "required when cryptographic binding is used"

**Review by paulbastian — APPROVED** [review]
I like this version. CNF is more optional to me and having required as the first keyword seems less right

**Comment by paulbastian** [comment]
No, I just mean SD-JWT VCs can be key bound, but they don't need to be. Therefore it is an optional feature.

**Comment by paulbastian** [comment]
I see, the difference here is that if key binding shall be used, then cnf must be used while the other formulation leaves more space?

**Comment by Sakurann** [comment]
If that's the intent, `required when cryptographic binding is used` sounds more appropriate to me. just saying `optional` would lead to confusion that it is optional even when cryptographic binding is used.

**Comment by awoie** [comment]
I also like the proposed version more. The proposed language addresses #196 which is based on developer feedback that were confused by the current language that uses REQUIRED if ... . 

So I think that OPTIONAL seems to solve that issue. IMO, it is also appropriate since we have the additional requirement in the paragraph that says MUST be present if cryptographic binding is required.

**Review by awoie — APPROVED** [review]
lgtm, see my comment in the discussion.

**Comment by paulbastian** [comment]
> > if key binding shall be used, then cnf must be used while the other formulation leaves more space?
> 
> Sorry, I don't understand why more space is needed here? The less optionality in key binding, more interop. with the current PR, it would be ok to use a top level sub claim and put a DID there (instead of using cnf.kid). I think it is harmful for the interop, that SD-JWT VC is hoped to bring

I see your point. The proposed text opens a door for key binding without using the cnf claim. However, the new text is more initiative and we could add this constraint as an additional sentence or in the section on key binding

**Comment by bc-pi** [comment]
> > > if key binding shall be used, then cnf must be used while the other formulation leaves more space?
> > 
> > 
> > Sorry, I don't understand why more space is needed here? The less optionality in key binding, more interop. with the current PR, it would be ok to use a top level sub claim and put a DID there (instead of using cnf.kid). I think it is harmful for the interop, that SD-JWT VC is hoped to bring
> 
> I don't see that this optionality is allowed by the following?
> 
> > This claim MUST be present when cryptographic Key Binding is to be supported.

**Comment by paulbastian** [comment]
Okay, it is already present. Sorry I didn't recheck the changes. Lgtm then

**Comment by bc-pi** [comment]
I'm struggling to understand the disagreement/argument here. But I'm trying...

The text for cnf in the PR has:

>  OPTIONAL. Contains the confirmation method identifying the proof of possession key as defined in [@!RFC7800]. This claim MUST be present when cryptographic Key Binding is to be supported. It is RECOMMENDED that this contains a JWK as defined in Section 3.2 of [@!RFC7800]. For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by the key identified in this claim.

which seems to me to clearly says that cnf is the required way to do key binding, if doing key binding. Especially the "This claim MUST be present when cryptographic Key Binding is to be supported" sentence. 

Would changing that sentence to "This claim is REQUIRED when cryptographic Key Binding is to be supported" help? Maybe I'm missing the point of contention but that and the current PR text and the current text in the draft all say the same thing - that cnf is the one and only way to do key binding in SD-JWT VC but b/c key binding isn't itself required, the claim also isn't required.

File: draft-ietf-oauth-sd-jwt-vc.md [hunk 1]
```diff
@@ -240,7 +240,7 @@ accepted before validating. See [@!RFC7519] for more information.
 Verifiable Credential is no longer valid. See [@!RFC7519] for more
 information.
 * `cnf`
-    * REQUIRED when cryptographic Key Binding is to be supported. Contains the confirmation method as defined in [@!RFC7800]. It is RECOMMENDED that this contains a JWK as defined in Section 3.2 of [@!RFC7800]. For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by the key identified in this claim.
+    * OPTIONAL. Contains the confirmation method identifying the proof of possession key as defined in [@!RFC7800]. This claim MUST be present when cryptographic Key Binding is to be supported. It is RECOMMENDED that this contains a JWK as defined in Section 3.2 of [@!RFC7800]. For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by the key identified in this claim.
```
Comment (bc-pi) [inline draft-ietf-oauth-sd-jwt-vc.md:243]: changing that one sentence to "This claim is REQUIRED when cryptographic Key Binding is to be supported"

```suggestion
    * OPTIONAL. Contains the confirmation method identifying the proof of possession key as defined in [@!RFC7800]. This claim is REQUIRED when cryptographic Key Binding is to be supported. It is RECOMMENDED that this contains a JWK as defined in Section 3.2 of [@!RFC7800]. For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by the key identified in this claim.
```

File: draft-ietf-oauth-sd-jwt-vc.md [hunk 2]
```diff
@@ -240,7 +240,7 @@ accepted before validating. See [@!RFC7519] for more information.
 Verifiable Credential is no longer valid. See [@!RFC7519] for more
 information.
 * `cnf`
-    * REQUIRED when cryptographic Key Binding is to be supported. Contains the confirmation method as defined in [@!RFC7800]. It is RECOMMENDED that this contains a JWK as defined in Section 3.2 of [@!RFC7800]. For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by the key identified in this claim.
+    * OPTIONAL. Contains the confirmation method identifying the proof of possession key as defined in [@!RFC7800]. This claim is REQUIRED when cryptographic Key Binding is to be supported. It is RECOMMENDED that this contains a JWK as defined in Section 3.2 of [@!RFC7800]. For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by the key identified in this claim.
```
Comment (danielfett) [inline draft-ietf-oauth-sd-jwt-vc.md:243]: _Technically,_ the REQUIRED contradicts the OPTIONAL. So in order to not trigger any RFC2119 nerds, we could use: 
```suggestion
    * OPTIONAL unless cryptographic Key Binding is to be supported, in which case it is REQUIRED. Contains the confirmation method identifying the proof of possession key as defined in [@!RFC7800]. It is RECOMMENDED that this contains a JWK as defined in Section 3.2 of [@!RFC7800]. For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by the key identified in this claim.
```

**Review by danielfett — APPROVED** [review]
I'm generally fine with the change, but I think I want to be pedantic here and resolve the contradiction.

**Comment by bc-pi** [comment]
> I'm generally fine with the change, but I think I want to be pedantic here and resolve the contradiction.

You're suggested pedantic wording is indeed better than what I'd had. Thanks. I've accepted it w/ c6b138036e035e6df2a2f983aa4cf8b183d39fc2

**Review by Sakurann — APPROVED** [review]
thank you!

**Pull request merged**